tools/telemetry-generator/src/handlers/stageTimingRetriever.ts

```diff
@@ -10,60 +10,90 @@ interface ParsedJob {
+ */
 module.exports = function handler(fileData, logger): void {
-	// - fileData is a JSON object obtained by calling JSON.parse() on the contents of a file.
-	// In this particular handler, we are using the timeline REST API to retrieve the status of the pipeline:
-	// Ex. https://dev.azure.com/fluidframework/internal/_apis/build/builds/<buildId>/timeline?api-version=6.0-preview.1
-	// - logger is an ITelemetryBufferedLogger. Call its send() method to write the output telemetry
-	//   events.
 	if (fileData.records?.length === undefined || fileData.records?.length === 0) {
-		console.log(`could not locate records info`);
+		throw new Error("No records found in the input data.");
 	}
 	if (process.env.BUILD_ID === undefined) {
-		console.log("BUILD_ID not defined.");
-	} else {
-		console.log("BUILD_ID", process.env.BUILD_ID);
+		throw new Error("BUILD_ID environment variable is not set.");
 	}
+	if (process.env.PIPELINE === undefined) {
+		throw new Error("PIPELINE environment variable is not set.");
+	}
+
+	console.log("BUILD_ID:", process.env.BUILD_ID);
+	console.log("PIPELINE:", process.env.PIPELINE);
+	console.log("STAGE_ID:", process.env.STAGE_ID);
 
-	// Note: type == "Task" would include tasks from the stages in the result set. It might be interesting in the future - for now we will only collect stages.
 	const parsedJobs: ParsedJob[] = fileData.records
-		.filter((job) => job.type === "Stage")
-		.map((job) => {
-			const startTime = Date.parse(job.startTime?.toString()) ?? undefined;
-			const finishTime = Date.parse(job.finishTime?.toString()) ?? undefined;
-			const dateDiff =
-				finishTime && startTime ? Math.abs(finishTime - startTime) / 1000 : undefined; // diff in seconds
-			console.log(`Name=${job.name}`);
+		// Note: type === "Task" or type === "Job" would include task-level (or job-level, respectively) telemetry.
+		// It might be interesting in the future - for now we will only collect stage-level telemetry.
+		// If given a specific STAGE_ID, only process that stage. Otherwise process all stages.
+		.filter((job) => job.type === "Stage" && (process.env.STAGE_ID === undefined || job.identifier === process.env.STAGE_ID))
+		.map((job): ParsedJob | undefined => {
+			console.log(
+				`Processing stage - name='${job.name}' identifier='${job.identifier}' state='${job.state}' result='${job.result}'`,
+			);
+
+			const finishTime = Date.parse(job.finishTime?.toString());
+			if (Number.isNaN(finishTime)) {
+				console.error(
+					`Failed to parse finishTime '${job.finishTime}'. The specified pipeline stage might not have finished yet. Telemetry for this stage will not be sent.`,
+				);
+				return undefined;
+			}
+
+			let startTime: number = finishTime;
+			if (job.state === "completed" && job.startTime === null) {
+				// A null start time when 'state === completed' indicates the stage was skipped.
+				// Set startTime to finishTime so duration ends up being 0.
+				startTime = finishTime;
+			} else {
```

Comment:
FWIW - this is already true. startTime is originally set to finishTime. Could change if and skip this block.

Reply:
I should have actually initialized it to -1, I prefer this check for skipped stages and why it makes sense to set `start = finish` to be explicit. I was thinking of the whole `if` block as the logic to assign `startTime` (or exit early). That said, I think I can do it more cleanly with a ternary assignment and the check to exit early after that, I'll do that.